### ISSUE TITLE
use after free in affile_dw_reap

### ISSUE THREAD

**Issue by hannob** [balabit/syslog-ng#2454]
## Version of syslog-ng

```
syslog-ng 3 (3.17.2)
Config version: 3.17
Installer-Version: 3.17.2
Revision: 
Compile-Date: Dec  7 2018 12:12:58
Module-Directory: /usr/lib64/syslog-ng
Module-Path: /usr/lib64/syslog-ng
Available-Modules: dbparser,date,map-value-pairs,cryptofuncs,csvparser,system-source,xml,tags-parser,afuser,stardate,cef,snmptrapd-parser,geoip-plugin,hook-commands,tfgetent,syslogformat,afsocket,linux-kmsg-format,json-plugin,kvformat,pseudofile,appmodel,add-contextual-data,disk-buffer,mod-python,graphite,confgen,basicfuncs,affile,afprog,afstomp
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: off
Enable-TCP-Wrapper: off
Enable-Linux-Caps: off
Enable-Systemd: off
```

## Platform

Linux

# Issue

There's a use after free bug in syslog-ng that can be detected with address sanitizer. This is likely a security issue.

## Failure

```
==7538==ERROR: AddressSanitizer: heap-use-after-free on address 0x612000007770 at pc 0x7fc3a89069c8 bp 0x7ffd8099afd0 sp 0x7ffd8099afc0
READ of size 8 at 0x612000007770 thread T0
    #0 0x7fc3a89069c7 in affile_dw_reap modules/affile/affile-dest.c:140
    #1 0x7fc3ac21f563 in iv_run_timers /var/tmp/portage/dev-libs/ivykis-0.42.3-r1/work/ivykis-0.42.3/src/iv_timer.c:119
    #2 0x7fc3ac22703f in iv_main /var/tmp/portage/dev-libs/ivykis-0.42.3-r1/work/ivykis-0.42.3/src/iv_main_posix.c:98
    #3 0x7fc3adf1e6d4 in main_loop_run lib/mainloop.c:580
    #4 0x401ef7 in main syslog-ng/main.c:307
    #5 0x7fc3ad45fb9d in __libc_start_main (/lib64/libc.so.6+0x21b9d)
    #6 0x4021b9 in _start (/usr/sbin/syslog-ng+0x4021b9)

```

## Steps to reproduce

1. compile syslog-ng with address sanitizer (-fsanitize=address in CFLAGS+LDFLAGS)

2. run syslog-ng, wait a while

**Comment by furiel** [comment]
Thanks for reporting but this is already fixed in https://github.com/balabit/syslog-ng/pull/2418.

**Comment by MrAnno** [comment]
The fix will be included in the syslog-ng OSE v3.19.1 release.
